fw/http_msg.c

```diff
@@ -1313,115 +1329,253 @@ tfw_http_msg_expand_data(TfwMsgIter *it, struct sk_buff **skb_head,
+						   &np);
 
-			return -ENOSPC;
-		}
-		cpy(mit->curr_ptr, c->data + c_off, n_copy);
+	if (unlikely(!addr))
+		return -ENOMEM;
 
-		T_DBG3("%s: acc_len=%lu, n_copy=%u, mit->curr_ptr='%.*s',"
-		       " ptr_diff=%ld\n", __func__, mit->acc_len, n_copy,
-		       n_copy, mit->curr_ptr, stop - mit->curr_ptr);
+	r = ss_skb_add_frag(it->skb_head, it->skb, addr, ++it->frag,
+			    FRAME_HEADER_SIZE);
+	if (unlikely(r))
+		return r;
 
-		mit->curr_ptr += n_copy;
-		mit->acc_len += n_copy;
+	ss_skb_adjust_data_len(mit->iter.skb, FRAME_HEADER_SIZE);
+	mit->frame_head = addr;
+	mit->curr_ptr = addr + FRAME_HEADER_SIZE;
 
-		/*
-		 * The fragment is exhausted. Move to the next fragment (or skb)
-		 * and reset the @mit->curr_ptr pointer.
-		 */
-		if (c_size >= f_room) {
-			if (skb_shinfo(it->skb)->nr_frags > it->frag + 1) {
-				skb_frag_t *frag;
+	return 0;
+}
 
-				++it->frag;
-				frag = &skb_shinfo(it->skb)->frags[it->frag];
-				mit->curr_ptr = skb_frag_address(frag);
-			}
-			else {
-				if (it->skb_head == it->skb->next
-				    && (c_size != f_room || c + 1 < end))
-					return -ENOSPC;
+/*
+ * Expand message by @str increasing size of current paged fragment or add
+ * new paged fragment using @pool if room in current pool's chunk is not enough.
+ */
+static int
+__tfw_http_msg_expand_from_pool(TfwHttpTransIter *mit, TfwPool* pool,
+				const TfwStr *str,
+				void cpy(void *dest, const void *src, size_t n))
+{
+	const TfwStr *c, *end;
+	unsigned int room;
+	unsigned int n_copy;
+	unsigned int rlen;
+	unsigned int off;
+	int r;
 
-				tfw_h2_msg_transform_setup(mit, it->skb->next,
-							   false);
-			}
+	BUG_ON(mit->iter.frag < 0);
 
-			if (c_size != f_room) {
-				c_off += n_copy;
-				goto this_chunk;
-			}
+	TFW_STR_FOR_EACH_CHUNK(c, str, end) {
+		rlen = c->len;
+
+		while (rlen) {
+			room = TFW_POOL_CHUNK_ROOM(pool);
+			BUG_ON(room < 0);
+
+			/*
+			 * Use available room in current pool chunk.
+			 * If pool chunk is exhausted new page will be allocated.
+			 */
+			n_copy = room == 0 ? rlen : min(room, rlen);
+			off = c->len - rlen;
+
+			r = tfw_http_msg_alloc_from_pool(mit, pool, n_copy);
+			if (unlikely(r))
+				return r;
+
+			cpy(mit->curr_ptr, c->data + off, n_copy);
+			rlen -= n_copy;
+			mit->acc_len += n_copy;
+
+			T_DBG3("%s: acc_len=%lu, n_copy=%u, mit->curr_ptr=%pK",
+			       __func__, mit->acc_len,
+			       n_copy, mit->curr_ptr);
 		}
 	}
 
 	return 0;
 }
 
 int
-tfw_h2_msg_rewrite_data(TfwHttpTransIter *mit, const TfwStr *str,
-			const char *stop)
+tfw_http_msg_expand_from_pool(TfwHttpTransIter *mit, TfwPool* pool,
+			      const TfwStr *str)
 {
-	return tfw_h2_msg_rewrite_data_common(mit, str, stop, memcpy_fast);
+	return __tfw_http_msg_expand_from_pool(mit, pool, str, memcpy_fast);
 }
 
 int
-tfw_h2_msg_rewrite_data_lc(TfwHttpTransIter *mit, const TfwStr *str,
-			   const char *stop)
+tfw_http_msg_expand_from_pool_lc(TfwHttpTransIter *mit, TfwPool* pool,
```

Comment:
`tfw_h2_msg_rewrite_data()` and `tfw_h2_msg_rewrite_data_lc()` should be removed from http_msg.h